packages/app-desktop/gui/note-viewer/index.html

```diff
@@ -195,7 +188,66 @@
 			return true;
 		}
 
-		let checkAllImageLoadedIID_ = null;
+		let alreadyAllImagesLoaded = false;
+
+		// During a note is being rendered, its height is varying. To keep scroll
+		// consistency, observing the height of the content element and updating its
+		// scroll position is required. For the purpose, 'ResizeObserver' is used.
+		// ResizeObserver is standard and an element's counterpart to 'window.resize'
+		// event. It's overhead is cheaper than observation using an interval timer.
+		//
+		// To observe the scroll height of the content element, adding, removing and
+		// resizing of its children should be observed. So, the combination of
+		// ResizeObserver (used for resizing) and MutationObserver (used for ading
+		// and removing) is used.
+		//
+		// References:
+		// https://developer.mozilla.org/en-US/docs/Web/API/ResizeObserver
+		// https://developer.mozilla.org/en-US/docs/Web/API/MutationObserver
+		//
+		// By using them, this observeRendering() function provides a efficient way
+		// to observe the changes of the scroll height of the content element
+		// using a callback approach.
+		function observeRendering(callback, compress = false) {
+			let previousHeight = 0;
+			const fn = (cause) => {
+				const height = contentElement.scrollHeight;
+				const heightChanged = height != previousHeight;
+				if (!compress || heightChanged) {
+					previousHeight = height;
+					callback(cause, height, heightChanged);
+				}
+			};
+			// 'resized' means DOM Layout change or Window resize event
+			let resizeObserver = new ResizeObserver(() => fn('resized'));
+			// An HTML document to be rendered is added and removed as a child of
+			// the content element for each setHtml() invocation.
+			let mutationObserver = new MutationObserver(entries => {
+				const e = entries[0];
+				e.removedNodes.forEach(n => n instanceof Element && resizeObserver.unobserve(n));
+				e.addedNodes.forEach(n => n instanceof Element && resizeObserver.observe(n));
+				if (e.removedNodes.length + e.addedNodes.length) fn('dom-changed');
+			});
+			mutationObserver.observe(contentElement, { childList: true });
+			return { mutationObserver, resizeObserver };
+		};
+
+		// A callback anonymous function invoked when the scroll height changes.
+		const onRendering = observeRendering((cause, height, heightChanged) => {
+			if (!alreadyAllImagesLoaded) {
+				const loaded = allImagesLoaded();
+				if (loaded) {
+					alreadyAllImagesLoaded = true;
+					ipcProxySendToHost('syncViewerScrollWithEditor', true);
+					ipcProxySendToHost('noteRenderComplete');
+					return;
+				}
+			}
+			if (heightChanged) {
+				// When the scroll height changes, sync is needed.
+				ipcProxySendToHost('syncViewerScrollWithEditor');
+			}
+		});
```

Comment:
Do you know if there's a performance impact on very large notes?

Also the notes sometimes are giant blobs of HTML, when it's imported from the web clipper or Evernote. Could you check that it works with these too (in particular that it doesn't freeze)?

Reply:
I also tested perf. The longest note was 61297 words with 3000 lines and I didn't notice a thing.

Reply:
I also tested a web-clipped page from [Long pages - Wikipedia](https://en.wikipedia.org/wiki/Special:LongPages). It consist of 2k lines and 60k words. There is no problem found.